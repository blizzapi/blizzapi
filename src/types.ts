/** Common */

export type Uri = string;

/** Generic */

export enum ConstantKey {
  REGIONS = "REGIONS",
  LOCALES = "LOCALES",
  DEFAULT_LOCALES = "DEFAULT_LOCALES",
  REGION_API_HOSTS = "REGION_API_HOSTS",
  SC2_REALMS = "SC2_REALMS",
  OAUTH_AUTHORIZE_URIS = "OAUTH_AUTHORIZE_URIS",
  OAUTH_TOKEN_URIS = "OAUTH_TOKEN_URIS",
  OAUTH_CHECK_TOKEN_URIS = "OAUTH_CHECK_TOKEN_URIS",
}

/** Regions */

export enum RegionName {
  us = "us",
  eu = "eu",
  kr = "kr",
  tw = "tw",
  cn = "cn",
}

export enum RegionId {
  us = 1,
  eu = 2,
  kr = 3,
  tw = 3,
  cn = 5,
}

export enum RegionIdAsString {
  us = "1",
  eu = "2",
  kr = "3",
  tw = "3",
  cn = "5",
}

export type RegionIdKey = string;
export type RegionIdAsNumberOrString = RegionId | RegionIdAsString;
export type RegionIdArray = ReadonlyArray<RegionId>;
export type RegionNameArray = ReadonlyArray<RegionName>;

/** Locales */

export enum Locale {
  en_US = "en_US",
  es_MX = "es_MX",
  pt_BR = "pt_BR",
  en_GB = "en_GB",
  es_ES = "es_ES",
  fr_FR = "fr_FR",
  ru_RU = "ru_RU",
  de_DE = "de_DE",
  pt_PT = "pt_PT",
  it_IT = "it_IT",
  ko_KR = "ko_KR",
  zh_TW = "zh_TW",
  zh_CN = "zh_CN",
}

export type LocaleArray = ReadonlyArray<Locale>;

/** Default locales */

export enum DefaultLocaleIndex {
  us = 0,
  eu = 0,
  kr = 0,
  tw = 0,
  cn = 0,
}

export type DefaultLocaleIndexArray = ReadonlyArray<DefaultLocaleIndex>;

/** StarCraft II realms  */

export enum Sc2Realm {
  US = 1,
  LatAm = 2,
  Europe = 1,
  Russia = 2,
  Korea = 1,
  Taiwan = 2,
  China = 1,
}

export enum Sc2RealmAsString {
  US = "1",
  LatAm = "2",
  Europe = "1",
  Russia = "2",
  Korea = "1",
  Taiwan = "2",
  China = "1",
}

export type Sc2RealmArray = ReadonlyArray<Sc2Realm>;
export type Sc2RealmList = ReadonlyArray<Sc2Realm>;
export type Sc2RealmAsNumberOrString = Sc2Realm | Sc2RealmAsString;

/** API class types */

export type RegionIdOrName = RegionId | RegionName;
export type MaybeRegion = RegionIdOrName | undefined;
export type ClientId = string;
export type ClientSecret = string;
export type AccessToken = string | undefined;
export type Endpoint = string;

export enum HttpMethod {
  GET = "GET",
  POST = "POST",
}

export type ValidatorFunction = (endpoint: Endpoint) => boolean;

/* OAuth flow URIs */

export enum OAuthHost {
  us = "https://us.battle.net",
  eu = "https://eu.battle.net",
  kr = "https://apac.battle.net",
  tw = "https://apac.battle.net",
  cn = "https://www.battlenet.com.cn",
}

export enum RegionHost {
  us = "https://us.api.blizzard.com",
  eu = "https://eu.api.blizzard.com",
  kr = "https://kr.api.blizzard.com",
  tw = "https://tw.api.blizzard.com",
  cn = "https://gateway.battlenet.com.cn",
}

export enum OAuthEndpoint {
  authorize = "/oauth/authorize",
  token = "/oauth/token",
  checkToken = "/oauth/check_token?token=",
}

/** API headers */

export enum ApiHeaders {
  LastModified = "last-modified",
}

/** Errors */

export enum ErrorCode {
  NotAuthorized = 401,
}

export enum ErrorResponseMessage {
  InvalidToken = "invalid_token",
  AccessTokenInvalid = "access_token_invalid",
  AccessTokenExpired = "access_token_invalid",
}

export type InvalidAccessTokenError = {
  error: ErrorResponseMessage.AccessTokenInvalid;
};

export type ExpiredAccessTokenError = {
  error: ErrorResponseMessage.AccessTokenExpired;
};

export type ResponseError = InvalidAccessTokenError | ExpiredAccessTokenError;

/** Interfaces */

export interface AccessTokenOptions {
  validateAccessTokenOnEachQuery?: boolean;
  refreshExpiredAccessToken?: boolean;
  onAccessTokenExpired?: () => void;
  onAccessTokenRefresh?: (newAccessToken: string) => void;
}

export interface QueryOptions {
  timeout?: number;
  headers?: object | Headers;
  params?: object | URLSearchParams;
}

export interface BattleNetOptions {
  region: RegionIdOrName;
  clientId: ClientId;
  clientSecret: ClientSecret;
  accessToken?: AccessToken;
}

export interface InitOptions {
  region: string | number;
  clientId?: string;
  clientSecret?: string;
  accessToken?: string;
  validateAccessTokenOnEachQuery?: boolean;
  revalidateAccessTokenIfExpired?: boolean;
  onAccessTokenExpire?: () => void;
  onAccessTokenRevalidate?: () => void;
}

export interface BattleNetQueryOptions {
  region: RegionIdOrName;
  endpoint: string;
  clientId: string;
  clientSecret: string;
  accessToken: AccessToken;
  options: AccessTokenOptions & QueryOptions;
}

export type RegionIdProperties<Value> = {
  [key in RegionId]: Value;
};

export interface BlizzAPIOptions extends BattleNetOptions, AccessTokenOptions {}
