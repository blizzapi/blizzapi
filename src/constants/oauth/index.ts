import { oAuthHosts } from "./OAuthHosts";
import { RegionIdProperties, RegionId, Uri } from "../../types";

const getRegionPropertyArray = (endpoint: string): RegionIdProperties<Uri> =>
  Object.freeze({
    [RegionId.us]: `${oAuthHosts[RegionId.us]}${endpoint}`,
    [RegionId.eu]: `${oAuthHosts[RegionId.eu]}${endpoint}`,
    [RegionId.kr]: `${oAuthHosts[RegionId.kr]}${endpoint}`,
    [RegionId.cn]: `${oAuthHosts[RegionId.cn]}${endpoint}`,
  });

const authorizeEndpoint = "/oauth/authorize";
const tokenEndpoint = "/oauth/token";
const checkTokenEndpoint = "/oauth/check_token?token=";

export const authorizeUris = getRegionPropertyArray(authorizeEndpoint);
export const tokenUris = getRegionPropertyArray(tokenEndpoint);
export const checkTokenUris = getRegionPropertyArray(checkTokenEndpoint);
