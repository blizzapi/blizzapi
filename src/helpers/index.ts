export * from "./bnet";
export * from "./fetch";
export * from "./oauth";
export * from "./validators";
