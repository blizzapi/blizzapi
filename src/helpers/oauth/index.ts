export { getAccessTokenObject } from "./getAccessTokenObject";
export { validateAccessToken } from "./validateAccessToken";
