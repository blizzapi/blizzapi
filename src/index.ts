import BlizzAPI from './lib/classes/BlizzAPI';
export = BlizzAPI;