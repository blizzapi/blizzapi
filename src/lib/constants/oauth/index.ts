import { RegionIdProperties } from '../../../../@interfaces';
import { Uri } from '../../../../@types';
import OAuthHosts from './OAuthHosts';

const getRegionPropertyArray = (endpoint: string): RegionIdProperties<Uri> => ({
  1: `${OAuthHosts[1]}${endpoint}`,
  2: `${OAuthHosts[2]}${endpoint}`,
  3: `${OAuthHosts[3]}${endpoint}`,
  5: `${OAuthHosts[5]}${endpoint}`,
});

const authorizeEndpoint = '/oauth/authorize';
const tokenEndpoint = '/oauth/token';
const checkTokenEndpoint = '/oauth/check_token?token=';

const authorizeUris = getRegionPropertyArray(authorizeEndpoint);
const tokenUris = getRegionPropertyArray(tokenEndpoint);
const checkTokenUris = getRegionPropertyArray(checkTokenEndpoint);

export default {
  authorizeUris: Object.freeze(authorizeUris),
  tokenUris: Object.freeze(tokenUris),
  checkTokenUris: Object.freeze(checkTokenUris),
}
