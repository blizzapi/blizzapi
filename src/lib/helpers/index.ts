export default {
  query: (endpoint: string) => {
    console.log(`query: ${endpoint}`);
  }
}