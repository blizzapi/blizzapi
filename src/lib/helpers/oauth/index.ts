export { default as fetchAccessToken } from './fetchAccessToken';
export { default as getAccessToken } from './getAccessToken';
export { default as validateAccessToken } from './validateAccessToken';
