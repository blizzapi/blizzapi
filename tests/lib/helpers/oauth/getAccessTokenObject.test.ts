import { getAccessTokenObject } from "../../../../src/helpers/oauth";
import { RegionName } from "../../../../src/types";

// TODO: Rewrite tests

describe("getAccessTokenObject()", () => {
  const region = "us" as RegionName;
  const clientId = "sample client id";
  const clientSecret = "sample client secret";

  it("should be defined", () => {
    expect(getAccessTokenObject).toBeDefined();
  });

  it("should be function", () => {
    expect(getAccessTokenObject).toBeTruthy();
    expect(typeof getAccessTokenObject).toBe("function");
  });

  it("returns correct response for GET request to a valid uri", async () => {
    expect.assertions(1);
    const response = await getAccessTokenObject({
      region,
      clientId,
      clientSecret,
    });
    expect(response).toMatchSnapshot();
  });

  it("throws an error if no client secret is provided", () => {
    expect.assertions(1);
    expect(async () => {
      await getAccessTokenObject({ region });
    }).rejects.toThrowError(
      "Cannot get access token because no Battle.net client id was provided"
    );
  });

  it("throws an error if no client secret is provided", () => {
    expect.assertions(1);
    expect(async () => {
      await getAccessTokenObject({ region, clientId });
    }).rejects.toThrowError(
      "Cannot get access token because no Battle.net client secret was provided"
    );
  });
});
