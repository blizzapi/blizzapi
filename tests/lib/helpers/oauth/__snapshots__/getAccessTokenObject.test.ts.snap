// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getAccessTokenObject() returns correct response for GET request to a valid uri 1`] = `
{
  "access_token": "no_access_token_provided",
  "headers": {
    "data": "default sample headers",
  },
  "params": {
    "data": "default sample params",
  },
  "timeout": 10000,
  "title": "sample mocked data for POST request to https://us.battle.net/oauth/token",
}
`;
